### PR TITLE
Add automatic tiebreaker for search requests that use a PIT

### LINKED ISSUE #56828: Virtual Sort field for automatic tie-breaking
The pagination of search requests using `search_after` require to use a tiebreaker that is unique per document. This is done automatically on sorted `_scroll` queries by tie-breaking documents on the index/shardId/docID tuple. This tuple is not accessible to normal search requests so the other option is to copy the `_id` of the document into a doc value field and use it as a tiebreaker. 
This solution is difficult to implement for solutions that are not in charge of indexation.
With the introduction of the [`search context`](https://github.com/elastic/elasticsearch/pull/56480) for requests, we'll be able to paginate over a set of sorted results using `search_after` with the guarantee to see the same documents during the walk. Since the internal document id wouldn't change between requests, using the tuple that `_scroll` queries use become possible. 
This issue proposes to expose a virtual sort field called `_tiebreak` (or any name that suits better). The field would be accessible as a sort criteria that can be used with a `search context` to ensure consistent ordering. The field would be composed of:
- The index UUID
- The shard ID
- The internal document ID

The order of the composition should be discussed but the main goal is to allow consistent ordering  using `search_after` without relying on manual operations at index-time.

### PR DESCRIPTION
**Pull request by jimczi**
This PR adds the special `_shard_doc` sort tiebreaker automatically to any
search requests that use a PIT. Adding the tiebreaker ensures that any
sorted query can be paginated consistently within a PIT.

Closes #56828

### CODE REVIEW

File: docs/reference/search/search-your-data/paginate-search-results.asciidoc [hunk 1]
```diff
@@ -90,18 +94,47 @@ GET /_search
...
-<2> Sorts hits for the search.
+<2> Sorts hits for the search with an implicit tiebreak on `_shard_doc` ascending.
 
 The search response includes an array of `sort` values for each hit. If you used
-a PIT, the response's `pit_id` parameter contains an updated PIT ID.
+a PIT, a tiebreakeris included as the last `sort` values for each hit.
+This tiebreaker called `_shard_doc` is added automically on every search requests that use a PIT.
+The `_shard_doc` value is the combination of the shard index within the PIT and the Lucene's internal doc ID,
+it is unique per document and constant within a PIT.
+You can also add the tiebreaker explicitely in the search request to customize the order:
```
Comment (mayya-sharipova) [inline docs/reference/search/search-your-data/paginate-search-results.asciidoc:111]: explicitely -> explicitly

File: server/src/main/java/org/elasticsearch/action/search/SearchRequest.java [hunk 2]
```diff
@@ -641,6 +644,31 @@ public int resolveTrackTotalHitsUpTo() {
...
+        }
+
+        SearchSourceBuilder source = this.source.rewrite(ctx);
+        boolean hasChanged = source != this.source;
+
+        // add a sort tiebreaker for PIT search requests if not explicitly set
+        if (source.pointInTimeBuilder() != null) {
+            if (source.sorts() == null || source.sorts().isEmpty()) {
+                source.sort(SortBuilders.scoreSort());
+            }
```
Comment (mayya-sharipova) [inline server/src/main/java/org/elasticsearch/action/search/SearchRequest.java:660]: I am not super clear why do we need to add sort on `_shard_doc` if the initial query was without sort at all?

Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/action/search/SearchRequest.java:660]: synchronized with @jimczi offline, and indeed we intend to add `_shard_doc` even for queries without any explicit sort. As these queries implicitly use sort on _score + shard + _doc. With this PR we make this sort obvious.

File: docs/reference/search/search-your-data/paginate-search-results.asciidoc [hunk 3]
```diff
@@ -71,9 +71,13 @@ To get the first page of results, submit a search request with a `sort`
 argument. If using a PIT, specify the PIT ID in the `pit.id` parameter and omit
 the target data stream or index from the request path.
 
-IMPORTANT: We recommend you include a tiebreaker field in your `sort`. This
-tiebreaker field should contain a unique value for each document. If you don't
-include a tiebreaker field, your paged results could miss or duplicate hits.
+IMPORTANT: If you cannot use a PIT, we recommend that you include a tiebreaker field
```
Comment (mayya-sharipova) [inline docs/reference/search/search-your-data/paginate-search-results.asciidoc:74]: may be we need to add  a more clear note:  All PIT search requests add an implicit tiebreak field called `_shard_doc`.

**Review by mayya-sharipova — APPROVED** [review]
@jimczi Thanks Jim, exciting change!
May be we can also add a note to the issue of possible future steps that we discussed offline:
1) don't add sort on _score for constant score queries
2) add sort on date desc for data streams etc.

File: docs/reference/search/search-your-data/paginate-search-results.asciidoc [hunk 4]
```diff
@@ -90,18 +96,47 @@ GET /_search
...
 ----
 // TEST[catch:missing]
 
 <1> PIT ID for the search.
-<2> Sorts hits for the search.
+<2> Sorts hits for the search with an implicit tiebreak on `_shard_doc` ascending.
 
 The search response includes an array of `sort` values for each hit. If you used
-a PIT, the response's `pit_id` parameter contains an updated PIT ID.
+a PIT, a tiebreakeris included as the last `sort` values for each hit.
```
Comment (astefan) [inline docs/reference/search/search-your-data/paginate-search-results.asciidoc:109]: Typo: `a tiebreakeris included` -> `a tiebreaker is included`

File: docs/reference/search/search-your-data/paginate-search-results.asciidoc (hunk 1 above)
Comment (jimczi) [reply docs/reference/search/search-your-data/paginate-search-results.asciidoc]: thanks

File: docs/reference/search/search-your-data/paginate-search-results.asciidoc (hunk 3 above)
Comment (jimczi) [reply docs/reference/search/search-your-data/paginate-search-results.asciidoc]: ++, I pushed https://github.com/elastic/elasticsearch/pull/68833/commits/2591295ea9a7a07a21b44ec26404d0ca129e0359

File: docs/reference/search/search-your-data/paginate-search-results.asciidoc (hunk 4 above)
Comment (jimczi) [reply docs/reference/search/search-your-data/paginate-search-results.asciidoc]: thanks

**Review by mayya-sharipova — APPROVED** [review]
@jimczi  Thanks for the new changes, this PR lgtm with the new changes as well

**Pull request merged**